examples/c++/userPropagator.cpp

```diff
@@ -0,0 +1,420 @@
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <random>
+#include <stack>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include <cstring>
+#include "z3++.h"
+
+/**
+ * The program solves the n-queens problem (number of solutions) with 4 different approaches
+ * 1) Bit-Vector constraints + Default solver + Blocking Clauses
+ * 2) Bit-Vector constraints + Simple solver + Blocking Clauses
+ * 3) Bit-Vector constraints + Simple solver + Adding contradictions in the propagator
+ * 4) Constraints only implicit via the propagator + Simple solver + Adding contradictions in the propagator
+ *
+ * Runs 1 + 2 are done for comparision with 3 + 4
+ */
+
+using namespace std::chrono;
+using std::to_string;
+
+#define QUEEN
+#define REPETITIONS 5
+
+#define SIZE(x) std::extent<decltype(x)>::value
+
+#if LOG
+#define WriteEmptyLine std::cout << std::endl
+#define WriteLine(x) std::cout << (x) << std::endl
+#define Write(x) std::cout << x
+#else
+#define WriteEmptyLine
+#define WriteLine(x)
+#define Write(x)
+#endif
+
+class model {
+
+public:
+
+    unsigned *values;
+    unsigned cnt;
+
+    model(unsigned cnt) : cnt(cnt), values(nullptr) {
+        values = (unsigned*) malloc(sizeof(unsigned) * cnt);
+    }
+
+    ~model() {
+        free(values);
+        values = nullptr;
+    }
+
+    unsigned &operator[](const int id) {
+        return values[id];
+    }
+
+    void set(unsigned pos, unsigned value) const {
+        values[pos] = value;
+    }
+
+    bool operator==(const model &other) const {
+        if (cnt != other.cnt) {
+            return false;
+        }
+        for (unsigned i = 0; i < cnt; i++) {
+            if (values[i] != other.values[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    bool operator!=(const model &other) const {
+        return !this->operator==(other);
+    }
+
+    model* clone() const {
+        model *newModel = new model(cnt);
+        memcpy(newModel->values, values, sizeof(unsigned) * cnt);
+        return newModel;
+    }
+};
+
+struct model_hash_function {
+    std::size_t operator()(const model &m) const {
+        size_t hash = 0;
+        for (unsigned i = 0; i < m.cnt; i++) {
+            hash *= m.cnt;
+            hash += m.values[i];
+        }
+        return hash;
+    }
+};
+
+class user_propagator : public z3::user_propagator_base {
+
+protected:
+
+    std::unordered_map<unsigned, unsigned> *id_mapping;
```

Comment:
I think you want & instead of *. It saves a character on dereference (. instead of ->) and captures invariant of being non-null